tokio/src/sync/mpsc/bounded.rs

```diff
@@ -1103,6 +1130,99 @@ impl<T> fmt::Debug for Sender<T> {
     }
 }
 
+impl Future for ReserveInner<'_> {
+    type Output = Result<(), SendError<()>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.project().acquire.poll(cx).map(|res| match res {
+            Ok(_) => Ok(()),
+            Err(_) => Err(SendError(())),
+        })
+    }
+}
+
+fn option_insert_pinned<T>(mut opt: Pin<&mut Option<T>>, value: T) -> Pin<&mut T> {
+    opt.as_mut().set(Some(value));
+    opt.as_pin_mut().unwrap()
+}
+
+impl<'a, T> Future for Reserve<'a, T> {
+    type Output = Result<Permit<'a, T>, SendError<()>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut me = self.project();
+        let inner = match me.inner.as_mut().as_pin_mut() {
+            Some(inner) => inner,
+            None => option_insert_pinned(me.inner, me.tx.reserve_inner()),
+        };
+        inner.poll(cx).map_ok(|()| Permit { chan: &me.tx.chan })
+    }
+}
+
+impl<T> fmt::Debug for Reserve<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Reserve").field("tx", &self.tx).finish()
+    }
+}
+
+impl<T> ReserveOwned<T> {
+    /// Aborts the reservation, returning the [`Sender`].
+    ///
+    /// After this method is called, this future is invalid in the same way as if it had been
+    /// polled to completion.
+    pub fn abort(self: Pin<&mut Self>) -> Sender<T> {
+        let mut me = self.project();
+        me.inner.set(None);
+        me.tx
+            .take()
+            .expect("ReserveOwned::abort called after completion")
+    }
+
+    /// Gets a reference to the [`Sender`] for the channel this future is reserving a permit in.
+    pub fn sender(&self) -> &Sender<T> {
+        self.tx
+            .as_ref()
+            .expect("ReserveOwned::sender called after completion")
+    }
+}
+
+impl<T> Future for ReserveOwned<T> {
+    type Output = Result<OwnedPermit<T>, SendError<()>>;
```

Comment:
Not that this can be changed without it being breaking, but I noticed this should probably return a `SendError<Sender<T>>`, like `try_reserve_owned()` does.